src/arch/io/network.cc

```diff
@@ -621,6 +621,430 @@ void linux_tcp_conn_t::on_event(int /* events */) {
+    void reclaim() {
+        conn = NULL;
+    }
+
+private:
+    SSL *conn;
+};
+
+/* This is the client version of the constructor. The base class constructor
+will establish a TCP connection to the peer at the given host:port and then we
+wrap the tcp connection in TLS using the configuration in the given tls_ctx. */
+linux_secure_tcp_conn_t::linux_secure_tcp_conn_t(SSL_CTX *tls_ctx, const ip_address_t &host, int port, signal_t *interruptor, int local_port) THROWS_ONLY(connect_failed_exc_t, interrupted_exc_t) :
+        linux_tcp_conn_t(host, port, interruptor, local_port) {
+
+
+    conn = SSL_new(tls_ctx);
+
+    /* If establish_conn() throws an exception, the conn_wrapper deallocates
+    the connection when unwinding the stack. */
+    tls_conn_wrapper_t conn_wrapper(conn);
+
+    establish_conn(SSL_connect);
+
+    conn_wrapper.reclaim();
+}
+
+/* This is the server version of the constructor */
+linux_secure_tcp_conn_t::linux_secure_tcp_conn_t(SSL_CTX *tls_ctx, fd_t sock) :
+        linux_tcp_conn_t(sock) {
+
+    conn = SSL_new(tls_ctx);
+
+    try {
+        establish_conn(SSL_accept);
+    } catch (const tcp_conn_t::connect_failed_exc_t &err) {
+        // Ignore.
+        logNTC("linux_secure_tcp_conn_t server constructor exception: %s\n", err.info.c_str());
+    }
+}
+
+linux_secure_tcp_conn_t::~linux_secure_tcp_conn_t() THROWS_NOTHING {
+    assert_thread();
+
+    if (is_open()) shutdown();
+
+    SSL_free(conn);
+}
+
+/* Creates a new OpenSSL connection using the underlying fd and the given
+OpenSSL context. The given handshake function is used to perform the handshake
+for the client or server using SSL_connect or SSL_accept respectively. */
+void linux_secure_tcp_conn_t::establish_conn(int (*handshake)(SSL *) ) THROWS_ONLY(connect_failed_exc_t) {
+    // Add support for partial writes.
+    SSL_set_mode(conn, SSL_MODE_ENABLE_PARTIAL_WRITE);
+
+    if (NULL == conn) {
+        unsigned long err_code = ERR_get_error();
+        throw linux_tcp_conn_t::connect_failed_exc_t(err_code, ERR_error_string(err_code, NULL));
+    }
+
+    // Set the underlying IO.
+    if (0 == SSL_set_fd(conn, sock.get())) {
+        unsigned long err_code = ERR_get_error();
+        throw linux_tcp_conn_t::connect_failed_exc_t(err_code, ERR_error_string(err_code, NULL));
+    }
+
+    // Perform TLS handshake.
+    while (true) {
+        ERR_clear_error();
+        int ret = handshake(conn);
+
+        if (ret > 0) {
+            return; // Successful TLS handshake.
+        }
+
+        int err_code = SSL_get_error(conn, ret);
+        unsigned long io_err = 0;
+
+        switch (err_code) {
+        case SSL_ERROR_WANT_READ:
+            /* There's no data available right now, so we must wait for a
+            notification from the epoll queue, or for an order to shut down. */
+            {
+                linux_event_watcher_t::watch_t watch(get_event_watcher(), poll_event_in);
+                wait_any_t waiter(&watch);
```

Comment:
The `wait_any_t` is not necessary here, because it's only waiting on `watch` anyway. You could call `wait_lazily_unordered()` directly on `watch`.

However we should also make sure that the handshake can get interrupted if for example the user shuts down the server while we're in here.
`establish_conn` should take the `signal_t *interruptor` as an argument, and here you should use
```C++
wait_interruptible(&watch, interruptor);
```

`wait_interruptible` throws an `interrupted_exc_t` if it gets interrupted (which also needs to be added to the `THROWS_ONLY` declaration of `establish_conn` above).